empress/miscs/input_generator.py

```diff
@@ -0,0 +1,161 @@
+import itertools
+import os
+import random
+
+from empress import input_reader
+random.seed(1)
+
+def generate_all_recon_input(n_leaves: int) -> input_reader.ReconInput:
+    """
+    Create an iterable of all possible ReconInput object with specified number of host/parasite leaves.
+    """
+    for host_dict in _generate_all_host_dict(n_leaves):
+        for parasite_dict in _generate_all_parasite_dict(n_leaves):
+            for mapping in _generate_all_tip_mapping(host_dict, parasite_dict):
+                yield input_reader.ReconInput(host_dict, None, parasite_dict, None, mapping)
+
+def generate_random_recon_input(n_leaves: int) -> input_reader.ReconInput:
+    """
+    Randomly create one ReconInput object with specified number of host/parasite leaves.
+    The randomization is not guaranteed to be uniform.
+    """
+    host_dict = _generate_random_host_dict(n_leaves)
+    parasite_dict = _generate_random_parasite_dict(n_leaves)
+    mapping = _generate_random_tip_mapping(host_dict, parasite_dict)
+    return input_reader.ReconInput(host_dict, None, parasite_dict, None, mapping)
+
+
+def _generate_all_host_dict(n_leaves):
+    yield from _generate_all_trees(n_leaves, 'hTop', 'h')
+
+def _generate_all_parasite_dict(n_leaves):
+    yield from _generate_all_trees(n_leaves, 'pTop', 'p')
+
+def _generate_all_trees(n_leaves: int, top_name: str, prefix: str):
+    """
+    We don't actually generate all possible trees in this
+    format. In the code, you will see that we only generate a right-heavy
+    tree, meaning that the right subtree has equal or greater number of
+    nodes than the left subtree. We do this to reduce the number of trees
+    that are isomorphic to one another.
+    """
+    tree_dict = {}
+    if n_leaves == 1:
+        tree_dict[top_name] = ('Top', prefix + '0', None, None)
```

Comment:
can we handle this case in generate_all_trees_helper by checking whether n_leaves is 0?

Reply:
I'm not sure I understand your comment. We only call the `_generate_all_trees_helper` if we have at least two leaves. Are you saying that even if we have one leaf, we should call both `_generate_all_trees_helper` on the left subtree and on the right subtree even if there is only one subtree? Would you mind clarifying how this should be modified?

Reply:
yeah, your understanding is correct. This is only a minor suggestion, the code works fine either way~

Reply:
I still don't understand your comment, and it does not seem that you are planning to explain it, so I will just merge.